specs/_features/peerdas/polynomial-commitments-sampling.md

```diff
@@ -0,0 +1,512 @@
+    for p in poly:
+        o.append(int(p) * factor_power % BLS_MODULUS)
+        factor_power = factor_power * inv_factor % BLS_MODULUS
+    return o
+```
+
+#### `interpolate_polynomialcoeff`
+
+```python
+def interpolate_polynomialcoeff(xs: Sequence[BLSFieldElement], ys: Sequence[BLSFieldElement]) -> PolynomialCoeff:
+    """
+    Lagrange interpolation: Finds the lowest degree polynomial that takes the value ``ys[i]`` at ``x[i]``
+    for all i.
+    Outputs a coefficient form polynomial. Leading coefficients may be zero.
+    """
+    assert len(xs) == len(ys)
+    r = [0]
+
+    for i in range(len(xs)):
+        summand = [ys[i]]
+        for j in range(len(ys)):
+            if j != i:
+                weight_adjustment = bls_modular_inverse(int(xs[i]) - int(xs[j]))
+                summand = multiply_polynomialcoeff(
+                    summand, [(- int(weight_adjustment) * int(xs[j])) % BLS_MODULUS, weight_adjustment]
+                )
+        r = add_polynomialcoeff(r, summand)
+    
+    return r
+```
+
+#### `vanishing_polynomialcoeff`
+
+```python
+def vanishing_polynomialcoeff(xs: Sequence[BLSFieldElement]) -> PolynomialCoeff:
+    """
+    Compute the vanishing polynomial on ``xs`` (in coefficient form)
+    """
+    p = [1]
+    for x in xs:
+        p = multiply_polynomialcoeff(p, [-int(x), 1])
+    return p
+```
+
+#### `evaluate_polynomialcoeff`
+
+```python
+def evaluate_polynomialcoeff(polynomial_coeff: PolynomialCoeff, z: BLSFieldElement) -> BLSFieldElement:
+    """
+    Evaluate a coefficient form polynomial at ``z`` using Horner's schema
+    """
+    y = 0
+    for coef in polynomial_coeff[::-1]:
+        y = (int(y) * int(z) + coef) % BLS_MODULUS
+    return BLSFieldElement(y % BLS_MODULUS)
+```
+
+### KZG multiproofs
+
+Extended KZG functions for multiproofs
+
+#### `compute_kzg_proof_multi_impl`
+
+```python
+def compute_kzg_proof_multi_impl(
+        polynomial_coeff: PolynomialCoeff,
+        zs: Sequence[BLSFieldElement]) -> Tuple[KZGProof, Sequence[BLSFieldElement]]:
+    """
+    Helper function that computes multi-evaluation KZG proofs.
+    """
+
+    # For all x_i, compute p(x_i) - p(z)
+    ys = [evaluate_polynomialcoeff(polynomial_coeff, z) for z in zs]
+    interpolation_polynomial = interpolate_polynomialcoeff(zs, ys)
+    polynomial_shifted = add_polynomialcoeff(polynomial_coeff, neg_polynomialcoeff(interpolation_polynomial))
+
+    # For all x_i, compute (x_i - z)
+    denominator_poly = vanishing_polynomialcoeff(zs)
+
+    # Compute the quotient polynomial directly in evaluation form
+    quotient_polynomial = divide_polynomialcoeff(polynomial_shifted, denominator_poly)
+
+    return KZGProof(g1_lincomb(KZG_SETUP_G1_MONOMIAL[:len(quotient_polynomial)], quotient_polynomial)), ys
+```
+
+#### `verify_kzg_proof_multi_impl`
+
+```python
+def verify_kzg_proof_multi_impl(commitment: KZGCommitment,
+                                zs: BLSFieldElement,
+                                ys: BLSFieldElement,
```

Comment:
Should `ys` be a `Sequence[BLSFieldElement]`?

```suggestion
                                ys: Sequence[BLSFieldElement],
```

Reply:
Oh I think you're right. Same for `xs` too, right?